### DISCUSSION TITLE
Why `control_in_blocking` for my UVC device returns `Err(Cancelled)` on Windows

### DISCUSSION THREAD

**Discussion by wangxiaochuTHU** [kevinmehall/nusb#43] [Q&A] [answered]
I think nusb is cool, so I'm willing to replace `rusb` with it in my project. However, in the very beginning, I got stuck with ***control read***.

I'm pretty sure that the following parameters work, with respect to the framework of rusb/libus. However, they failed with nusb.
```rust
let request = 0x86; 
let wvalue = 0x0100; 
let windex = 0x01; 
let timeout = std::time::Duration::from_secs(1);
let mut buf = [0; 50];
```

**rusb** `read_control` works
```rust
/* with rusb, these parameters work */
let size = device_handle
        .read_control(request_type, request, wvalue, windex, &mut buf, timeout)
        .unwrap();  
```

**nusb** `control_in_blocking` fails and returns `Err(Cancelled)`.
```rust
/* with nusb, these parameters fails*/
let control = Control {
        control_type: ControlType::Class,
        recipient: Recipient::Interface,
        request: request,
        value: wvalue,
        index: windex,
    };
let result = interface.control_in_blocking(control, &mut buf, timeout);
```

I've replaced the driver of my usb device to `WinUSB` by Zadig, and I also claimed the interface before using it. 

Could you give me some guidlines to get over it? Thank you very much.

## log
```
 2024-03-01T14:00:11.2461282+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Probing device "USB\\VID_04B4&PID_00F9\\5&2FCC3441&0&24"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:36

  2024-03-01T14:00:11.2463681+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Probing device "USB\\VID_413C&PID_2113\\5&2FCC3441&0&9"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:36

  2024-03-01T14:00:11.2466523+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Probing device "USB\\VID_0403&PID_6001\\A601OO93"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:36

  2024-03-01T14:00:11.2467724+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Probing device "USB\\VID_09DA&PID_C10A\\5&2FCC3441&0&8"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:36

  2024-03-01T14:00:11.2471259+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Probing device "USB\\VID_04B4&PID_00F9\\5&2FCC3441&0&24"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:36

  2024-03-01T14:00:11.2472468+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Probing device "USB\\VID_413C&PID_2113\\5&2FCC3441&0&9"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:36

  2024-03-01T14:00:11.2474764+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Probing device "USB\\VID_0403&PID_6001\\A601OO93"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:36

  2024-03-01T14:00:11.2475957+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Probing device "USB\\VID_09DA&PID_C10A\\5&2FCC3441&0&8"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:36

  2024-03-01T14:00:11.2477167+08:00 DEBUG nusb::platform::windows_winusb::device: Creating device for "USB\\VID_04B4&PID_00F9\\5&2FCC3441&0&24"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\device.rs:43

  2024-03-01T14:00:11.2478107+08:00 DEBUG nusb::platform::windows_winusb::enumeration: Driver is "WinUSB"
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\enumeration.rs:122

  2024-03-01T14:00:11.2480071+08:00  INFO nusb::platform::windows_winusb::device: Blocking control In, 50 bytes
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\device.rs:165

  2024-03-01T14:00:12.2562724+08:00 ERROR nusb::platform::windows_winusb::device: WinUsb_ControlTransfer failed: The semaphore timeout period has expired (os error 121)
    at C:\Users\dell\.cargo\registry\src\mirrors.ustc.edu.cn-61ef6e0cd06fb9b8\nusb-0.1.6\src\platform\windows_winusb\device.rs:204
```
# Updated

## `DeviceInfo` of my usb device  
```rust
Configuration {
    configuration_value: 1,
    num_interfaces: 2,
    attributes: 128,
    max_power: 50,
    string_index: None,
    interface_alt_settings: [
        InterfaceAltSetting {
            interface_number: 0,
            alternate_setting: 0,
            num_endpoints: 1,
            class: 14,
            subclass: 1,
            protocol: 0,
            string_index: Some(
                2,
            ),
            endpoints: [
                Endpoint {
                    address: 0x82,
                    direction: In,
                    transfer_type: Interrupt,
                    max_packet_size: 64,
                    packets_per_microframe: 1,
                    interval: 1,
                },
            ],
        },
        InterfaceAltSetting {
            interface_number: 1,
            alternate_setting: 0,
            num_endpoints: 1,
            class: 14,
            subclass: 2,
            protocol: 0,
            string_index: None,
            endpoints: [
                Endpoint {
                    address: 0x83,
                    direction: In,
                    transfer_type: Bulk,
                    max_packet_size: 1024,
                    packets_per_microframe: 1,
                    interval: 0,
                },
            ],
        },
    ],
}
```
```

**Comment by kevinmehall** [comment]
The log `WinUsb_ControlTransfer failed: The semaphore timeout period has expired (os error 121)` indicates that it's reaching the timeout. There is 1 second elapsed since the prior log entry. Currently timeouts get `TransferError::Cancelled`.

It's hard to tell why that would be different than rusb, though.

One limitation to be aware of is that with `Recipient::Interface`, WinUSB will force `index` to the interface number that you passed to `claim_interface`. So if your interface number is not 1, the device may be receiving a different `index` than you think, and if it doesn't handle that error properly, maybe it doesn't respond with a `STALL` and just times out. The same would apply to rusb / libusb, though, if it's also using the WinUSB driver.

**Reply by wangxiaochuTHU** [reply]
Yes, my usb device has 2 interfaces indeed. I think `index = 0` would be what the device actually wants, and I will try it out later. 

Thanks for the very accurate guideline.

# Update
After testing, it seems sad that `index = 0` didn't solve my problem. I also iterated `index` (from 0 to 2 ) to test which set of parameters could work but none did, i.e.
```rust
for control_type in [ControlType::Class, ControlType::Standard, ControlType::Vendor ] {
        for recipient in [ Recipient::Device, Recipient::Endpoint, Recipient::Interface, Recipient::Other ] {
            for index in 0..=2 {
                if let Ok(_) = wrap_nusb_control_in_blocking(control_type, recipient, &device, &interface, index) {
                    panic!()
                }
            }
        }
    }
//   <----- can reach here
```

Thanks anyway. I can't figure the problem out till now. Maybe I have to investigate it by debugging into the device's hardware code later. But please tell me if you have any suggestions based on the device information updated in the question.

**Comment by wangxiaochuTHU** [comment]
Hi! After investigated, my problem appears to be related to two low-level issues. 

1. Firmware-level
My firmware treats that transfer (as posted above) sent via `libusb` as a `stream request`, while treats the one sent via `WinUSB` as a `control request`. This difference then results in two different handing branches in the firmware.

2. Transport-level
The `index`-field of struct `Control` in `nusb`  appears to ***LOSE LOWER BYTE*** during the transport, e.g. set `index` to 0x0102 at nusb side, but after transport, that seen value at the firmware is not 0x0102.

Would you consider resolving these two issues at `nusb`'s level?

**Reply by kevinmehall** [reply]
> My firmware treats that transfer (as posted above) sent via libusb as a stream request, while treats the one sent via WinUSB as a control request. This difference then results in two different handing branches in the firmware.

I'm not sure what you mean by "stream request". Do you mean different bmRequestType values resulting leading to different branches in the setup packet handler as I discuss below? With nusb, the bmRequestType is calculated from the `control_type` and `recipient` fields and the in or out direction.

> The index-field of struct Control in nusb appears to LOSE LOWER BYTE during the transport, e.g. set index to 0x0102 at nusb side, but after transport, that seen value at the firmware is not 0x0102.

This is the limitation I was referring to previously. In https://github.com/kevinmehall/nusb/pull/45 I added documentation and log a warning when this occurs. Note that this also applies with libusb and rusb if you are using the WinUSB driver and pass a `bmRequestType` with the recipient bits set to interface (though libusb can use other kernel drivers like libusb0 or libusbK without this limitation, but nusb uses WinUSB only). What driver were you using successfully with rusb, and what `request_type` value were you passing? 

Are you sure you want `Recipient::Interface` and not `Recipient::Device`? In rusb, the recipient is set by the 4 lowest bits of the `request_type`, but you don't show what `request_type` value you were using with rusb.

It's part of the USB standard that the the `index` least significant byte is supposed to contain the interface number when the recipient is set to Interface, so it is strange that you are passing index `0x0102` when there is no interface 2.

If you control the firmware, you could make it use `Recipient::Device` instead of `Recipient::Interface` to be able to use any `index` without problems.

**Comment by wangxiaochuTHU** [comment]
> I'm not sure what you mean by "stream request". Do you mean different bmRequestType values resulting leading to different branches in the setup packet handler as I discuss below? With nusb, the bmRequestType is calculated from the `control_type` and `recipient` fields and the in or out direction.

Emm... I think Yes, I mean the UVC-specific request can be detected by firmware, when using `libusb-win32 (libusb0)` driver.

> It's part of the USB standard that the the `index` least significant byte is supposed to contain the interface number when the recipient is set to Interface, so it is strange that you are passing index `0x0102` when there is no interface 2.

Yes you're right, and you've mentioned it before. But I wonder if it is essentially related to issue 1 (UVC-specific request). But seems that I took an inappropriate example, and I'd like to have taken another example `0x0001`.
> If you control the firmware, you could make it use `Recipient::Device` instead of `Recipient::Interface` to be able to use any `index` without problems.

Yes, I can control the firmware and change its inner logic.

---
Actually, I'm not expert in USB standard but a user at best, so I'm not sure what is its should-be state, and I'd appreciate the discussion and the guideline you've provided. 

But what made me most confused is the interoperability. 

In fact, for developping the firmware of the device (UVC device), several common video played (e.g. the OS out-of-the-box video player, *PotPlayer*, and *ffmpeg*) were used to test and check the interoperability. The device firmware has passed the check. Note that in this stage, there's no programmed software nor user-installed driver. Only default uvc drivers/frameworks (such as MSMF, DirectShow, V4L2) were used. So I assume that, after this stage, the firmware logic is compatible with uvc standard.

Then later, I focused on Windows and found that the OS framework (MSMF and DirectShow) seemed to be less performant since frames would drop/miss every hundreds of frames. So I turned to the way of reading raw data of usb and decoding w.r.t uvc. To achieve this, I installed `libusb-win32` for the device to replace the os-default-uvc driver, and I used `rusb` Rust crate to communicate with it. This way works at last.

After these stages, the uvc device turns out to be has interoperability with all these common players and libusb-win32 driver. So I feel strange why it would fail with WinUSB. Doesn't the Windows-default uvc driver not depend on Windows-default usb driver WinUSB? 

However, although I assume the firmware should better keep its current state to keep interoperability, I still tested modifying its inner logic to accept any `index` and installing WinUSB as the driver. In this case, `rusb` succeeds in capture the video, while `nusb` only succeeds in control transfer but it fails in reading bulk_in_blocking streams (CompErr(Unknown)). 

```rust
DEBUG nusb::platform::windows_winusb::enumeration: Driver is "WinUSB"
INFO nusb::platform::windows_winusb::device: Blocking control In, 50 bytes
INFO nusb::platform::windows_winusb::device: Blocking control In, 50 bytes
INFO nusb::platform::windows_winusb::device: Blocking control In, 50 bytes
INFO nusb::platform::windows_winusb::device: Blocking control Out, 34 bytes
INFO nusb::platform::windows_winusb::device: Blocking control In, 50 bytes
INFO nusb::platform::windows_winusb::device: Blocking control Out, 34 bytes
DEBUG nusb::platform::windows_winusb::transfer: Submit transfer 0x1985f385790 on endpoint 83 for 102400 bytes IN
DEBUG nusb::platform::windows_winusb::transfer: Submit transfer 0x1985f385610 on endpoint 83 for 102400 bytes IN
DEBUG nusb::platform::windows_winusb::transfer: Handling completion for transfer 0x1985f385790
DEBUG nusb::platform::windows_winusb::transfer: Submit transfer 0x1985f3857f0 on endpoint 83 for 102400 bytes IN
DEBUG nusb::platform::windows_winusb::transfer: Handling completion for transfer 0x1985f385610
DEBUG nusb::platform::windows_winusb::transfer: Transfer 0x1985f385790 on endpoint 83 failed: 87, 0 bytes transferred
```

At last, although `rusb` work with the `WinUSB` driver, the performance seems worse (more image drops) than with the `libusb0` driver . Is this comparison a common conclusion, i.e. WinUSB acts not good as libusb0, even if `WinUSB` runnings in the kernel space?

**Reply by kevinmehall** [reply, marked as answer]
Ah, ok, didn't realize that it was UVC. By "control" and "stream" you were also referring to the `VideoControl` and `VideoStreaming` interfaces defined in the UVC specification. In your descriptor you posted above, your interface number 0 has subclass=1 so is the VideoControl interface, and your interface number 1 has subclass=2 so is the VideoStreaming interface. 

You should `claim_interface(0)` and `claim_interface(1)`  to open both of these interfaces, and make `VideoControl` requests using interface 0, and `VideoStreaming` requests using interface 1. Then WinUSB will set the least significant byte of index to 0 or 1 to match the interface number used.

I think UVC will have an interface association descriptor that groups together those two interfaces. That makes Windows see them as a single interface group, and as a result, you might have a problem with nusb `claim_interface` on interface 1. WinUSB requires [WinUSB_GetAssociatedInterface](https://learn.microsoft.com/en-us/windows/win32/api/winusb/nf-winusb-winusb_getassociatedinterface) to open interfaces other than the first one within an interface association descriptor group, which nusb doesn't do right now, and libusb didn't handle this either until very recently.

If your whole device (rather than specific interface) is bound to WinUSB, https://github.com/kevinmehall/nusb/pull/41 may help, because in nusb 0.1.6 it was only looking at the first interface in that case. I'll release 0.1.7 with that fix soon.

Yes, since it is UVC, you should follow the standard, though since you control the firmware you could e.g. make it accept either recipient=Device OR recipient=Interface if you needed to.

The Windows UVC kernel driver is not built on WinUSB. WinUSB is a kernel driver whose purpose is to expose the device functionality to userspace for a driver written as a regular application instead of in kernel mode. I'd expect the kernel UVC driver to be faster than doing it in userspace with nusb or rusb, and you'd generally want to use the kernel mode driver for classes like UVC where one exists.

Typically UVC uses isochronous endpoints, which are not currently implemented in nusb (though are supported in WinUSB as of Windows 10). Are your endpoints bulk or isochronous? Isochronous might prevent you from using nusb for now.

If it is bulk, and since it looks like you are submitting 3+ transfers (which is the correct way to optimize performance) the performance difference between libusb0 and WinUSB drivers might be due to `RAW_IO`. The WinUSB `RAW_IO` option is essential for high speed streaming without dropping data.  nusb is designed with `RAW_IO` in mind, but [doesn't actually turn it on yet](https://github.com/kevinmehall/nusb/issues/6). libusb doesn't support RAW_IO and rusb doesn't expose the libusb async API needed to submit multiple concurrent transfers to take advantage of it.

**Reply by wangxiaochuTHU** [reply]
Thank you very much. My understanding goes deeper along with your knowledge.

--

>Typically UVC uses isochronous endpoints, which are not currently implemented in nusb (though are supported in WinUSB as of Windows 10). Are your endpoints bulk or isochronous? Isochronous might prevent you from using nusb for now.

Yes, so my uvc device uses bulk transfer other than isochronous to seek for wider compatibility.

I just tried to `claim_interface(0)` and `claim_interface(1)` simultaneously but encountered an error `Os { code: 5, kind: PermissionDenied, message: "Access is denied." } `. If with only ONE interface claimed,  the request recognized by uvc device would be always `VideoStreaming` (for either interface 0 or 1). Particularly, with interface No.1 claimed, the least significant byte of `index` still seemed to be overrode to 0. With a modified firmware that accepts both `VideoStreaming`/`VideoControl` and all `index`, bulk-in transfers still failed. All the above was tested with nusb version `0.1.7`. Observation also revealed that the speed of device was printed as `high speed` although it has been connected to a usb 3.0 port.

>The WinUSB RAW_IO option is essential for high speed streaming without dropping data. nusb is designed with RAW_IO in mind, but https://github.com/kevinmehall/nusb/issues/6. libusb doesn't support RAW_IO and rusb doesn't expose the libusb async API needed to submit multiple concurrent transfers to take advantage of it.

It sounds great! I expect one day this `RAW_IO` feature can thoroughly solve my use concerns. Thanks again.

**Reply by kevinmehall** [reply]
> I just tried to claim_interface(0) and claim_interface(1) simultaneously but encountered an error Os { code: 5, kind: PermissionDenied, message: "Access is denied." } . 

Yeah, https://github.com/kevinmehall/nusb/pull/41 doesn't actually fix this like I hoped it would. Opened https://github.com/kevinmehall/nusb/issues/48 for this.
